### PR TITLE
[megatron] support mixtral training with megatron backend 

### PR DESCRIPTION
**Pull request by ccclyu**
# What does this PR do?

Add support of Mixtral MOE model training with Megatron backend including ``Mixtral8x7B`` and ``Mixtral8X22B``. 

# ChangeLog:

It is still labor-heavy to add new type of model to ``mcore`` format including the following changes:
- ``hf_to_mcore_config_mixtral``: convert `hf_config` to `TransformerConfig`. some common configs are merged into one function `_get_base_transformer_config`. 
- ``MixtralModel`` in model_initialzier.py: implement a model initializer class to initialize GPTModel from config.
- `McoreToHFWeightConverterMixtral`:  model conversion class from mcore to huggingface basically rename
- model entry in `registry.py`: add entry function or class in corresponding registries.  

# Usage

- convert [Mixtral-8x7B-Instruct-v0.1](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1/) to mcore format ([converted](https://huggingface.co/clyu/Mixtral-8x7B-Instruct-v0.1-mcore/tree/main))
- Run RLOO script as follows:

```bash
set -x

train_files=$gsm8k_train_path
test_files=$gsm8k_test_path

export MEGATRON_MODEL="Mixtral-8x7B-Instruct-v0.1-mcore"

python3 -m verl.trainer.main_ppo --config-path=./config --config-name='ppo_megatron_trainer' \
    algorithm.adv_estimator=rloo \
    data.train_files=$train_files \
    data.val_files=$test_files \
    data.train_batch_size=128 \
    data.truncation="left" \
    data.max_prompt_length=512 \
    data.max_response_length=4096 \
    actor_rollout_ref.model.path=Mixtral-8x7B-Instruct-v0.1 \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=128 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.actor.megatron.tensor_model_parallel_size=8 \
    actor_rollout_ref.actor.megatron.pipeline_model_parallel_size=2 \
    actor_rollout_ref.actor.megatron.use_dist_checkpointing=True \
    actor_rollout_ref.actor.megatron.dist_checkpointing_path=${MEGATRON_MODEL} \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=8 \
    actor_rollout_ref.rollout.max_num_batched_tokens=8192 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.4 \
    actor_rollout_ref.rollout.n=4 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=128 \
    actor_rollout_ref.ref.megatron.tensor_model_parallel_size=8 \
    actor_rollout_ref.ref.megatron.pipeline_model_parallel_size=2 \
    actor_rollout_ref.ref.megatron.use_dist_checkpointing=True \
    actor_rollout_ref.ref.megatron.dist_checkpointing_path=${MEGATRON_MODEL} \
    algorithm.use_kl_in_reward=True \
    algorithm.kl_ctrl.kl_coef=0.001 \
    trainer.critic_warmup=0 \
    trainer.val_before_train=True \
    trainer.logger=['console','wandb'] \
    trainer.log_val_generations=100 \
    trainer.project_name='verl_gsm8k_test' \
    trainer.experiment_name='mixtral-8x7b-rloo-gsm8k' \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=2 \
    trainer.save_freq=50 \
    trainer.test_freq=10 \
    trainer.total_epochs=15
```
- RLOO training can improve GSM8K validation accuracy from 74% to 85%

# What is Missing

- refactor hf2mcore conversion scripts as https://github.com/volcengine/verl/pull/1267
- Have a good design of onboarding new model class to avoid labor-intension changes.  

## Before submitting

- [x] Did you read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide) and finish the [code format check](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting)? 
- [x] Did you make sure to update the documentations with your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs) especially for breaking config etc?
- [ ] Did you write any test cases if necessary? Please add CI tests to your new feature.  

# Additional Info: 
- **Issue Number**: None
- **Training**:  Megatron
- **Inference**:  None

### CODE REVIEW

**Comment by BearBiscuit05** [comment]
Great PR! Please complete the CLA first. Thanks!

**Review by eric-haibin-lin — COMMENTED** [review]
please add a record in https://github.com/volcengine/verl/blob/main/docs/experiment/ppo.rst

**Comment by ccclyu** [comment]
@ann-qin-lu could you also please review the code? thanks!

File: verl/models/mcore/model_forward.py [hunk 1]
```diff
@@ -18,7 +19,8 @@
 from .util import postprocess_packed_seqs, preprocess_packed_seqs, recover_left_padding, remove_left_padding
 
 
-def gptmodel_forward_dense(model, input_ids, attention_mask, position_ids, sequence_parallel, value_model=False, pack_seqs=True):
+def gptmodel_forward(model, input_ids, attention_mask, position_ids, sequence_parallel, value_model=False, pack_seqs=True):
```
Comment (dataproblems) [inline verl/models/mcore/model_forward.py:22]: does it make sense to create a model_type enum and pass that as an arg here to then determine which forward pass method to use depending on the model type?

File: verl/models/mcore/model_initializer.py [hunk 2]
```diff
@@ -14,132 +15,117 @@
...
+        self.tfconfig = tfconfig
+        self.hf_config = hf_config
+
+    @abstractmethod
+    def get_transformer_layer_spec(self):
+        """Get the transformer layer specification."""
+        pass
+
+    def get_rope_scaling_args(self):
+        rope_scaling_args = {}
```
Comment (dataproblems) [inline verl/models/mcore/model_initializer.py:41]: do we see a usecase where there would be additional fields in the rope scaling args?  If not, do you think we could return a single variable i.e. the  `self.hf_config.rope_scaling["factor"]` instead of a dictionary? Perhaps an Optional[T] instead of Optional[Dict[str, Any]].

File: verl/models/mcore/model_initializer.py [hunk 3]
```diff
@@ -14,132 +15,117 @@
...
+        if "rope_scaling" in self.hf_config:
+            if self.hf_config.rope_scaling is not None:
+                assert self.hf_config.rope_scaling["type"] == "linear", "only linear scaling is supported for now"
+                rope_scaling_args["seq_len_interpolation_factor"] = self.hf_config.rope_scaling["factor"]
+        return rope_scaling_args
+
+    def initialize(
```
Comment (dataproblems) [inline verl/models/mcore/model_initializer.py:48]: nit: Can we add a docstring that provides a brief summary and reference to the GPTModel class? ( not a blocker to release )

File: verl/models/mcore/model_initializer.py [hunk 4]
```diff
@@ -14,132 +15,117 @@
...
-    use_te = True
-    if freeze_moe_router:
-        tfconfig.moe_router_load_balancing_type = "none"
-
-    def patch_layer_spec(transformer_layer_spec):
-        # shared_experts.gate=True
+from .config_converter import PretrainedConfig, TransformerConfig
+
+
+class BaseModelInitializer(ABC):
```
Comment (dataproblems) [inline verl/models/mcore/model_initializer.py:26]: nit: Could you decorate with return types for the functions as well?  ( not a blocker for release )

File: verl/models/mcore/model_initializer.py [hunk 5]
```diff
@@ -14,132 +15,117 @@
...
+                layer.mlp.router.weight.requires_grad = False
+                layer.mlp.shared_experts.gate_weight.requires_grad = False
+        return model
+
+
+class MixtralModel(BaseModelInitializer):
+    """Initializer for Mixtral models."""
+
+    def get_transformer_layer_spec(self):
+        assert self.tfconfig.normalization == "RMSNorm", "only RMSNorm is supported for now"
```
Comment (dataproblems) [inline verl/models/mcore/model_initializer.py:127]: do you want to create a private method that just validates the tfconfig during initialization?

File: verl/models/mcore/registry.py [hunk 6]
```diff
@@ -22,35 +28,70 @@
...
 )
-from .weight_converter import McoreToHFWeightConverterDense, McoreToHFWeightConverterQwen2Moe
+from .weight_converter import (
+    McoreToHFWeightConverterDense,
+    McoreToHFWeightConverterMixtral,
+    McoreToHFWeightConverterQwen2Moe,
+)
+
+# Registry for model configuration converters
+MODEL_CONFIG_CONVERTER_REGISTRY: Dict[str, Callable[[PretrainedConfig, torch.dtype], TransformerConfig]] = {
```
Comment (dataproblems) [inline verl/models/mcore/registry.py]: Do you want to use enums for the keys here to avoid typos?

File: verl/models/mcore/registry.py [hunk 7]
```diff
@@ -59,51 +100,56 @@ def hf_to_mcore_config(hf_config: PretrainedConfig, dtype: torch.dtype) -> Trans
...
     return MODEL_FORWARD_REGISTRY[arch]
 
 
-def get_mcore_weight_converter(hf_config: PretrainedConfig, dtype: torch.dtype):
-    MODEL_WEIGHT_CONVERTER_REGISTRY = {
-        "LlamaForCausalLM": McoreToHFWeightConverterDense,
-        "Qwen2ForCausalLM": McoreToHFWeightConverterDense,
-        "Qwen2MoeForCausalLM": McoreToHFWeightConverterQwen2Moe,
-    }
+def get_mcore_weight_converter(hf_config: PretrainedConfig, dtype: torch.dtype) -> Any:
```
Comment (dataproblems) [inline verl/models/mcore/registry.py]: seems like the return value is a Callable, let's use that over Any.

File: verl/utils/vllm_utils.py [hunk 8]
```diff
@@ -42,11 +30,27 @@ def patch_vllm_moe_model_weight_loader(model):
...
+    # Define MLP attribute mapping for different model types
+    MLP_ATTR_MAPPING = {
+        MixtralForCausalLM: "block_sparse_moe",
+    }
+
+    # Default MLP attribute name for models not in the mapping
+    DEFAULT_MLP_ATTR = "mlp"
+
+    if not isinstance(model, SUPPORTED_MOE_MODELS):
         return
```
Comment (dataproblems) [inline verl/utils/vllm_utils.py:53]: let's add a log statement here indicating that the model is not an instance of the supported moe models to help users with understanding the reason for an empty return.

File: verl/models/mcore/registry.py (hunk 6 above)
Comment (ccclyu) [reply verl/models/mcore/registry.py]: good suggestion. yes, will use one `enums` for supportedmodel.

File: verl/models/mcore/registry.py (hunk 7 above)
Comment (ccclyu) [reply verl/models/mcore/registry.py]: thanks. fixed.

File: verl/models/mcore/config_converter.py [hunk 9]
```diff
@@ -15,86 +16,93 @@
...
 from transformers import PretrainedConfig
 
 
-def hf_to_mcore_config_dense(hf_config: PretrainedConfig, dtype: torch.dtype) -> TransformerConfig:
-    # for LlamaForCausalLM or Qwen2ForCausalLM
+def _get_base_transformer_config(hf_config: PretrainedConfig, dtype: torch.dtype, **kwargs) -> TransformerConfig:
+    """
+    Create a base TransformerConfig with common parameters across different model architectures.
+    TODO: (ycl) use dataclass or converter config?
+    https://github.com/NVIDIA/NeMo/blob/0af8b7df793ad7538f149ad9bcb8c2cae5134c1a/nemo/collections/llm/gpt/model/qwen2.py
```
Comment (ann-qin-lu) [inline verl/models/mcore/config_converter.py]: no need to add this link here?

File: verl/models/mcore/config_converter.py [hunk 10]
```diff
@@ -15,86 +16,93 @@
...
+        "pipeline_model_parallel_size": mpu.get_pipeline_model_parallel_world_size(),
+        "virtual_pipeline_model_parallel_size": mpu.get_virtual_pipeline_model_parallel_world_size(),
+        "context_parallel_size": mpu.get_context_parallel_world_size(),
+        "overlap_p2p_comm": overlap_p2p_comm,
+        "batch_p2p_comm": batch_p2p_comm,
+        "sequence_parallel": mpu.get_tensor_model_parallel_world_size() > 1,
+        # Common settings
+        "variable_seq_lengths": True,
+        "masked_softmax_fusion": True,
+        "moe_token_dispatcher_type": "alltoall",
```
Comment (ann-qin-lu) [inline verl/models/mcore/config_converter.py:74]: this is for MOE model arch? can we move to moe model section.

File: verl/models/mcore/config_converter.py [hunk 11]
```diff
@@ -15,86 +16,93 @@
...
+
+
+def hf_to_mcore_config_dense(hf_config: PretrainedConfig, dtype: torch.dtype) -> TransformerConfig:
+    # for LlamaForCausalLM or Qwen2ForCausalLM
+    qkv_bias = True if "Qwen2ForCausalLM" in hf_config.architectures else getattr(hf_config, "attention_bias", False)
+
+    return _get_base_transformer_config(
+        hf_config=hf_config,
+        dtype=dtype,
         use_cpu_initialization=True,
```
Comment (ann-qin-lu) [inline verl/models/mcore/config_converter.py]: why this is set to True, while the line 101 is set as False?

File: verl/models/mcore/config_converter.py [hunk 12]
```diff
@@ -15,86 +16,93 @@
...
+        # Common settings
+        "variable_seq_lengths": True,
+        "masked_softmax_fusion": True,
+        "moe_token_dispatcher_type": "alltoall",
+    }
+
+    # Update with any provided overrides
+    base_config.update(kwargs)
+
+    return TransformerConfig(**base_config)
```
Comment (ann-qin-lu) [inline verl/models/mcore/config_converter.py:81]: [nit] good to add a debug log here to show the overridden TFConfig.

File: verl/models/mcore/model_forward.py (hunk 1 above)
Comment (ann-qin-lu) [reply verl/models/mcore/model_forward.py:22]: nit: add inline comment to show what models have been tested with this method

File: verl/models/mcore/saver.py [hunk 13]
```diff
@@ -453,3 +453,7 @@ def _broadcast_tp_shard_tensor_qkv(tensor, q_name, k_name, v_name, src_pp_rank):
 
 def merge_megatron_ckpt_gptmodel_qwen_moe(wrapped_models, config, dtype, is_value_model=False, tie_word_embeddings=False):
     raise NotImplementedError("merge_megatron_ckpt_gptmodel_qwen_moe is not implemented")
+
+
+def merge_megatron_ckpt_gptmodel_mixtral(wrapped_models, config, dtype, is_value_model=False, tie_word_embeddings=False):
+    raise NotImplementedError("merge_megatron_ckpt_gptmodel_mixtral is not implemented")
```
Comment (ann-qin-lu) [inline verl/models/mcore/saver.py:459]: curious what's the blocker for implementing this function?

File: verl/models/mcore/config_converter.py (hunk 11 above)
Comment (ccclyu) [reply verl/models/mcore/config_converter.py]: will change to `False` as default.

File: verl/models/mcore/config_converter.py [hunk 14]
```diff
@@ -15,86 +16,93 @@
...
-        hidden_size=hf_config.hidden_size,
-        num_attention_heads=hf_config.num_attention_heads,
-        num_query_groups=hf_config.num_key_value_heads,
-        ffn_hidden_size=hf_config.intermediate_size,
-        activation_func=F.silu,
-        normalization="RMSNorm",
-        gated_linear_unit=True,
+
+    # Base configuration with common parameters
+    base_config = {
```
Comment (ETOgaosion) [inline verl/models/mcore/config_converter.py:46]: Current refactor is OK, and there is another way of better organizing these configs as Megatron arguments, using network/TE/training config to better enabling MCore performance improvement features.

File: verl/models/mcore/model_forward.py (hunk 1 above)
Comment (ccclyu) [reply verl/models/mcore/model_forward.py:22]: done.

File: verl/models/mcore/model_initializer.py (hunk 2 above)
Comment (ccclyu) [reply verl/models/mcore/model_initializer.py:41]: there are more `rope_scaling` args that can be extended in the future. like https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_rope_utils.py

File: verl/models/mcore/config_converter.py (hunk 14 above)
Comment (ccclyu) [reply verl/models/mcore/config_converter.py:46]: sure. will re-design config logic in the future PR.

**Comment by ccclyu** [comment]
addressed comments from @ann-qin-lu @dataproblems and added mixtral 8x22b instruct results ([wandb](https://api.wandb.ai/links/ppo_dev/sbuiuf2d)) in doc @eric-haibin-lin . Can we now approve&merge? @ETOgaosion  Will refactor the code as proposed in https://github.com/volcengine/verl/issues/1371

File: verl/models/mcore/config_converter.py (hunk 10 above)
Comment (ccclyu) [reply verl/models/mcore/config_converter.py:74]: https://github.com/NVIDIA/Megatron-LM/blob/10b5c5898e950a4ecc62974ed834163713868921/megatron/core/transformer/transformer_config.py#L1058C1-L1063C18
```

        if self.moe_token_dispatcher_type in ['allgather', 'alltoall_seq']:
            if self.variable_seq_lengths is True:
                raise ValueError(
                    f"Token dispatcher type: {self.moe_token_dispatcher_type} does not support "
                    f"variable sequence length, please use alltoall dispatcher instead."
                )
```

the default value for `moe_token_dispatcher_type` is `allgather`, which causes the above conflict. we set `alltoall` explicitly here.
